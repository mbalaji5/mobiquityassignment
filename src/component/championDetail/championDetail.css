.championDetail {
  width: 700px;
  border-left: 10px solid #00afef;
  padding: 10px;
  box-sizing: content-box;
  text-align: left;
  background: #cfd7d7;
}
