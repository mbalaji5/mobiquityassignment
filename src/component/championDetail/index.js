export { default as ChampionDetail } from "./championDetail";
