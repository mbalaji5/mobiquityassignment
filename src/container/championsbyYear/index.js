export { default as ChampionsByYear } from "./championsByYear";
