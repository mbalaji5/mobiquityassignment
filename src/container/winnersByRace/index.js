export { default as WinnersByRace } from "./winnersByRace";
