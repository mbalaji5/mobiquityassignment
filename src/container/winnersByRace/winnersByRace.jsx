import React, { useEffect, useState } from "react";
import axios from "axios";
import { Error } from "../../component/error";
import { Loading } from "../../component/loading";
import { RaceDetail } from "../../component/raceDetail";

import PropTypes from "prop-types";
// Container that loads ist of the winners for every race for the selected year.
// Used functional component.
// useEffect and useState to handle the life cycle event.
// Should render error screen when there is error from api.
// Loading componet wil be displayed until the data is getting retrived.
// props-> selected Season and world champion id of selected season
// Async await is another option to load the data asynchronous.

const WinnersByRace = ({ season, championDriverId }) => {
  const [RacesData, setRacesData] = useState([]);
  const [isError, setIsError] = useState(false);
  const [isLoading, setIsLoading] = useState(true);
  useEffect(() => {
    axios
      .get(`http://ergast.com/api/f1/${season}/results/1.json?limit=100`)
      .then((data) => {
        const {
          data: {
            MRData: {
              RaceTable: { Races }
            }
          }
        } = data;

        setRacesData(Races);
        setIsLoading(false);
      })
      .catch((error) => {
        setIsError(true);
      });
  }, [season]);

  return (
    <div>
      {isLoading && <Loading />}
      {isError && !isLoading && <Error />}
      {RacesData &&
        !isLoading &&
        !isError &&
        RacesData.map((item) => {
          return <RaceDetail {...item} championDriverId={championDriverId} />;
        })}
    </div>
  );
};

WinnersByRace.propTypes = {
  season: PropTypes.string.isRequired,
  championDriverId: PropTypes.string.isRequired
};

WinnersByRace.defaultProps = {
  season: "2021",
  championDriverId: "driver1"
};
export default WinnersByRace;
