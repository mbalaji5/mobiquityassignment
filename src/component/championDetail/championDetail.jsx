import React, { useState } from "react";
import PropTypes from "prop-types";
import { WinnersByRace } from "../../container/winnersByRace";
import "./championDetail.css";

// Component render the Season,world champion name and other details
// Used functional component.
// useState to handle the state
// Should render WinnersByRace component when user clicks on specific season
// props-> list of world champions details and season

const ChampionDetail = ({ DriverStandings, season }) => {
  const [isShowWinnerByRace, setisShowWinnerByRace] = useState(false);

  const {
    Driver: { driverId, familyName, givenName },
    points,
    wins
  } = DriverStandings[0];

  const expandRace = () => {
    setisShowWinnerByRace(!isShowWinnerByRace);
  };
  return (
    <div className="championDetail" key={season} onClick={expandRace}>
      <h2>
        {season} - {familyName} {givenName}
      </h2>
      <span>
        Points:{points} | Wins:{wins}
      </span>
      {isShowWinnerByRace && (
        <WinnersByRace season={season} championDriverId={driverId} />
      )}
    </div>
  );
};

ChampionDetail.propTypes = {
  season: PropTypes.string.isRequired,
  DriverStandings: PropTypes.array.isRequired
};

ChampionDetail.defaultProps = {
  season: 2021,
  DriverStandings: []
};
export default ChampionDetail;
