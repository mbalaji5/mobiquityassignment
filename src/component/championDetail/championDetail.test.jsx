import React from "react";
import { render, screen } from "@testing-library/react";
import { ChampionDetail } from "../championDetail";
import { shallow } from "enzyme";
import { WinnersByRace } from "../../container/winnersByRace";

const props = {
  season: "2021",
  DriverStandings: [
    {
      Driver: {
        dateOfBirth: "1952-08-17",
        driverId: "piquet",
        familyName: "Piquet",
        givenName: "Nelson",
        nationality: "Brazilian",
        url: "http://en.wikipedia.org/wiki/Nelson_Piquet"
      },
      points: "73",
      position: "1",
      positionText: "1",
      wins: "3"
    }
  ]
};

describe("Should render champion detail", () => {
  it("Should validate season and name", () => {
    render(<ChampionDetail {...props} />);
    const element = screen.getByText(/2021 - Piquet Nelson/i);
    expect(element).toBeDefined();
  });
  it("Should validate point ans wins", () => {
    render(<ChampionDetail {...props} />);
    const element = screen.getByText(/Points:73 | Wins:3/i);
    expect(element).toBeDefined();
  });

  it("Should validate onclick", () => {
    const wrapper = shallow(<ChampionDetail {...props} />);

    wrapper
      .find(".championDetail")
      .props()
      .onClick();

    expect(wrapper.find(WinnersByRace)).toBeTruthy();
  });
  it("Should validate child param when onclick", () => {
    const wrapper = shallow(<ChampionDetail {...props} />);
    wrapper
      .find(".championDetail")
      .props()
      .onClick();
    const raceDetails = wrapper.find(WinnersByRace);
    expect(raceDetails).toBeDefined();
    expect(raceDetails.props().season).toBe(props.season);
    expect(raceDetails.props().championDriverId).toBe(
      props.DriverStandings[0].Driver.driverId
    );
  });
});
describe("Should validate champion detail", () => {
  it("Should validate shopshot", () => {
    const { baseElement } = render(<ChampionDetail {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
