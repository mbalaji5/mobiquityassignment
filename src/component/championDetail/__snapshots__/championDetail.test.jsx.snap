// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render champion detail shopshot Should validate shopshot 1`] = `
<body>
  <div>
    <div
      class="championDetail"
    >
      <h2>
        2021
         - 
        Piquet
         
        Nelson
      </h2>
      <span>
        Points:
        73
         | Wins:
        3
      </span>
    </div>
  </div>
</body>
`;

exports[`Should validate champion detail Should validate shopshot 1`] = `
<body>
  <div>
    <div
      class="championDetail"
    >
      <h2>
        2021
         - 
        Piquet
         
        Nelson
      </h2>
      <span>
        Points:
        73
         | Wins:
        3
      </span>
    </div>
  </div>
</body>
`;
